#9 updates readme, generates basic running app

willbuck opened the pull request:
Hey JHipster team!

I now have `mhipster` generating an app that runs successfully and can access all of our available admin functions (user admin, config, health, metrics, logs).

I stripped the entities from the https://github.com/jhipster/jhipster-sample-app-micronaut as the best I can understand, those are meant to be added by the `entity` generator, stitched into the app via this `needle-api`. Is that correct? 

If so, I'll have to do some iterating on that to ensure it works as well (I haven't done any work there though so my guess is the server portion of generating will not work currently.)

I also have not (yet) done any explicit overriding of the `prompts` to limit options based on what is currently supported (basically the choices in the `.yo-rc.json` of the sample app linked above). I can follow-up with some work on that, as I think it would be good UX to not allow prospective MHipsters to select options with no support.

Let me know your thoughts, thanks!

atomfrede commented on generators/server/templates/src/main/java/package/repository/AuthorityRepository.java.ejs:
```diff
@@ -0,0 +1,13 @@
+import io.micronaut.data.jpa.repository.JpaRepository;
+
+/**
+ * Micronaut Predator repository for the {@link Authority} entity.
```
Shouldn't we stick to the "official" name Micronaut Data?

willbuck replied:
Good catch! Fixed in latest PR

atomfrede commented on generators/server/templates/src/main/java/package/repository/UserRepository.java.ejs:
```diff
@@ -0,0 +1,63 @@
+import java.time.Instant;
+
+/**
+ * Spring Data JPA repository for the {@link User} entity.
```
```suggestion
 * Micronaut Data JPA repository for the {@link User} entity.
```

willbuck replied:
TY! Fixed that

atomfrede commented on generators/server/templates/src/main/java/package/security/SecurityUtils.java.ejs:
```diff
@@ -0,0 +1,59 @@
+import java.util.Optional;
+
+/**
+ * Utility class for Spring Security.
```
Is `Spring` here intended? Maybe confusing

willbuck replied:
No, you're right; fixed that, thanks for finding!

willbuck commented:
Looks like I have a few failing client tests for the `admin/config` module; I'll fix those up post haste

willbuck commented:
I've noticed now that the issue is a linting issue on the generator itself, not the project which is output (which I had been primarily concerned with). I will fix the linting errors tomorrow (currently trying to squash some test failures in the generated micronaut app: I'm missing some email templates and config (I think, WIP)

atomfrede commented:
:rocket: Huge work @willbuck Hard to check all files, but from what I see it looks fine. The frontend files are needed because of the different monitoring & configuration responses of micronaut, right? 
I would say we can merge it and evolve the supported options and extend ci a little if possible together. Maybe we can have now smaller issues to it is more transparent what can be worked on.

willbuck commented:
@atomfrede that's correct, and yeah my apologies for the giant dump of a PR... getting to a baseline of "generates a project that runs" should hopefully mean we can get a lot more granular with future improvements.

I have a few more backend tests to fix but I can fast-follow with that if we'd like to just merge this now; it does at least generate a working project for development!

atomfrede commented:
> I have a few more backend tests to fix but I can fast-follow with that if we'd like to just merge this now; it does at least generate a working project for development!

If you want to commit them first go ahead, no need to rush anything

sendilkumarn commented on generators/app/index.js:
```diff
@@ -0,0 +1,147 @@
+    //  * @return {String} The name of the root generator
+    //  */
+    // rootGeneratorName() {
+    //     return jhipsterPackagejs.name;
```
This is good to have to re-use certain generator properties from the JHipster generator. For now, this should be okay.

sendilkumarn commented on generators/app/index.js:
```diff
@@ -0,0 +1,147 @@
+                this.log(chalk.white('Welcome to MHipster  🤓  :: Running Micronaut version 1.3.0'));
+                this.log(chalk.white('This blueprint generates your backend as a Micronaut Java project.'));
+                this.log(chalk.green(' _______________________________________________________________________________________________________________\n'));
+                this.log(chalk.white(`  For any questions or improvements refer to the stream lead at ${chalk.yellow('https://github.com/willbuck')}`));
```
might be good to have a link back to the https://github.com/jhipster/generator-jhipster-micronaut repo

willbuck replied:
I believe there is one on the next line, no?

```
                this.log(
                    chalk.white(
                        `  If you find MHipster useful, support and star the project at ${chalk.yellow(
                            'https://github.com/jhipster/generator-jhipster-micronaut'
                        )}`
                    )
                );
```

willbuck replied:
Also if it is inappropriate to name myself as the stream lead I'm happy to remove / edit that line!

sendilkumarn commented on generators/client/files.js:
```diff
@@ -0,0 +1,151 @@
+const jhipsterConstants = require('generator-jhipster/generators/generator-constants');
```
Why are we overwriting the client generator. We should use the Client generator as such from the JHipster or else MHipster will become difficult to maintain.

willbuck replied:
Unfortunately need to overwrite some of the files to be compatible w/ the standard output from micronaut, as we are not a 100% match on top of what Spring Boot outputs (nor should we really be; for example, it wouldn't make sense to match the nested structure of Spring Boot's `config` endpoint when our config is treated as a flat map). 

Our `health`, `metrics`, and auth endpoints also have slightly different output, thus I had to make some small tweaks to the frontend for all of those things.

I had a similar feeling to you at first but after speaking with @jameskleeh about it we both agreed this was the right course of action. I agree that it adds some maintenance weight, but hopefully by only overriding a few of the admin modules, I've minimized the impact as best as it can be minimized. I'm open to more suggestions on how to improve it though!

sendilkumarn replied:
This makes sense.

sendilkumarn commented on generators/client/files.js:
```diff
@@ -0,0 +1,151 @@
+const clientFiles = {
+    angularMain: [
+        {
+            condition: generator => generator.clientFramework === 'angularX',
```
The files object is exported from the base-generator we can use that here rather than re-writing the files here

willbuck replied:
I'm not sure how you mean, can you elaborate?

sendilkumarn commented on generators/common/index.js:
```diff
@@ -0,0 +1,41 @@
+        const initMicronautCommonPhaseSteps = {};
+        return Object.assign(initPhaseFromJHipster, initMicronautCommonPhaseSteps);
+
+        // Here we are not overriding this phase and hence its being handled by JHipster
```
can we remove this if we are not using this?

sendilkumarn commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "generator-jhipster-micronaut",
-  "version": "0.0.0",
+  "version": "0.0.1",
```
Do not bump this yet, we will do this when we are releasing :)

willbuck replied:
Got it, will change, ty :)

willbuck commented:
@sendilkumarn @atomfrede  let me know if I've addressed all your concerns or if you feel there's still more to be done. I've fixed the tests so that the project now:

1. runs in dev mode via `./mvnw` + `npm start`, as a normal jhipster app would
2. accepts any `packageName` and `baseName` (had previously overlooked some spots, that is now fixed)
3. passes all tests via `./mvnw clean verify`
4. Can be packaged for use as a fat jar via `./mvnw package`

Right now this is my next list of priorities:
1. Solving a few latent issues (seems the prod config isn't *quite* right, getting a driver issue, also thymeleaf translation support wasn't working in the tests when I would think it ought to)
2. Limiting the prompts so users aren't able to select things that aren't supported
3. Entity insertion support
4. Gradle support
5. React support
6.  something else (other DBs, auth types, source langs etc)

Still, I think this would be a good point to merge so that PRs can start to look smaller.

sendilkumarn commented:
Thanks @willbuck 👏 

The builds are failing. can we fix it?

can we also remove travis from the project and use azure pipelines ?

sendilkumarn commented:
Created project board let us track and map the items that we want in the first release. Feel free to add things into it.

willbuck commented:
@sendilkumarn I can definitely take out the travis config, when you say "use azure pipelines" I'm unfamiliar w/ what that entails. I expect it would involve adding something similar to [this file](https://github.com/jhipster/generator-jhipster/blob/master/azure-pipelines.yml), though I would likely be paring down what's tested in that one to something more barebones. Does that gel with your understanding?

sendilkumarn commented:
Yeah in the main generator we had flaky pipeline using Travis, the Azure pipeline and GitHub Workflow is fast and more stable. So we can switch to them. But if you are more comfortable with Travis, we can live with that for now.

willbuck commented:
@sendilkumarn I have no personal preference between the tools (we use github actions on many micronaut projects already, and are migrating to it where we were previously using travis), I just don't necessarily know how best to orchestrate the new tool. It looks like it currently uses [a bunch of scripts](https://github.com/jhipster/generator-jhipster/tree/master/test-integration/scripts), and also has [github actions](https://github.com/jhipster/generator-jhipster/tree/master/.github/workflows). The kotlin project also does similar things w/ [scripts](https://github.com/jhipster/jhipster-kotlin/tree/master/test-integration/scripts), [azure](https://github.com/jhipster/jhipster-kotlin/blob/master/azure-pipelines.yml), AND [gh-workflows](https://github.com/jhipster/jhipster-kotlin/tree/master/.github/workflows), the node project doesn't use azure but [does use gh-workflows](https://github.com/jhipster/generator-jhipster-nodejs/tree/master/.github/workflows), the vue project [also uses gh-workflows](https://github.com/jhipster/jhipster-vuejs/tree/master/.github/workflows), and the dotnetcore is [still using travis](https://github.com/jhipster/jhipster-dotnetcore/blob/master/.travis.yml). 

It'd help me to have a little guidance around what the jhipster folks thinks *should* be covered here, as well as clarify how gh-workflows are currently being used vs these azure pipelines, and where / how I should focus my effort there (or whether it makes more sense to have one of the jhipster folks do the initial work of migrating).

To me, we should be:
1. Running the linter and tests for the generator itself on PR commits
2. Running a project generation and checking basic functionality (front and backend generate successfully, can start up okay, login works, entity generation works, entity CRUD functionality works in the running app) runs on PR commits.

I'm going to focus my effort today on adding the JIB/docker support stuff.

willbuck commented:
Okay @sendilkumarn I've got the docker support in there.

Is there anything else you (or the other jhipster folks) see as blocking this PR so that we can have a solid baseline to merge commits moving forward?

atomfrede commented:
@willbuck Huge work, amazing! I would suggest we merge this or is there anything you would like to add to this PR? The pipelines can be added separately and all other features too. As @jdubois said let's focus on sql support and maybe oauth support, I would agree that many other options are nice to have.

sendilkumarn commented:
Let us progressively decide the build tool. This is huge work. I would say merge this and then we will decide further. We will focus and make small PRs in future :)

willbuck commented:
@atomfrede Thank you for the kind words! Just so there's clarity, this currently supports JWT, but we do have Oauth2 support we can add in a later PR. See issue #16 

@sendilkumarn I agree that we'll get smaller PRs going forward and I think this is good to merge (also think the #12 PR is ready to follow fast for merging, as combined the two allow generating the sample app as-is)
